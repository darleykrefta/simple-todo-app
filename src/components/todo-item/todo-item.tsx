import React, { useCallback } from 'react'

import { useRecoilState } from 'recoil'

import { DeleteIcon } from '@chakra-ui/icons'
import { Box, Flex, IconButton, Text } from '@chakra-ui/react'

import { todoItemsStore } from 'stores/todo-store'

interface Props {
  id: string | number
  description: string
  completed: boolean
  handleDelete?: (id: string | number) => void
}

export const TodoItem: React.FC<Props> = ({ id, description, completed, handleDelete }) => {
  const [todoList, setTodoList] = useRecoilState(todoItemsStore)

  const handleCompleteItem = useCallback(() => {
    const itemIndex = todoList.findIndex(item => item.id === id)
    const todoItemUpdated = { ...todoList[itemIndex], completed: !completed }
    const todoItemsUpdated = [...todoList.slice(0, itemIndex), todoItemUpdated, ...todoList.slice(itemIndex + 1)]
    setTodoList(todoItemsUpdated)
  }, [id, completed, todoList, setTodoList])

  return (
    <Flex flex={1} justifyContent="space-between" alignItems="center">
      <Box
        p={4}
        m={1}
        w="300px"
        bg="secondary"
        border="2px"
        borderRadius="15px"
        borderColor="primary"
        cursor="pointer"
        onClick={() => handleCompleteItem()}
        data-testid="todo-container"
      >
        <Text style={{ ...(completed ? { textDecoration: 'line-through' } : {}) }} data-testid="todo-text">
          {description}
        </Text>
      </Box>

      {handleDelete ? (
        <IconButton
          bg="primary"
          color="white"
          _hover={{ bg: 'primary' }}
          borderRadius="50%"
          aria-label="delete-todo"
          onClick={() => handleDelete(id)}
          icon={<DeleteIcon />}
          data-testid="input-todo-delete"
        />
      ) : null}
    </Flex>
  )
}
