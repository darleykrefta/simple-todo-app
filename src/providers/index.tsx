import React from 'react'

import { RecoilRoot } from 'recoil'

import { ChakraProvider, CSSReset, extendTheme } from '@chakra-ui/react'

const theme = extendTheme({
  colors: {
    primary: '#590202',
    secondary: '#F26D3D',
    background: '#84BF9E'
  }
})

export const Providers: React.FC = ({ children }) => {
  return (
    <ChakraProvider theme={theme}>
      <RecoilRoot>
        <CSSReset />
        {children}
      </RecoilRoot>
    </ChakraProvider>
  )
}
