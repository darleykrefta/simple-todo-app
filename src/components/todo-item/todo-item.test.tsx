import React from 'react'

import { cleanup, fireEvent, render, screen, waitFor } from '@testing-library/react'

import { Providers } from 'providers'
import { TodoItem } from './todo-item'

interface Props {
  id: string | number
  description: string
  completed: boolean
  handleDelete?: (id: string | number) => void
}

const initialValuesMock: Props = { id: 'item', description: 'todo 1', completed: false }

const setup = (props: Props) => render(<TodoItem {...props} />, { wrapper: Providers })

describe('TodoList component', () => {
  afterEach(cleanup)

  it('should be valid with last version', () => {
    const { asFragment } = setup(initialValuesMock)
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render with initialValues correctly', () => {
    setup(initialValuesMock)
    const text = screen.getByTestId('todo-text')
    expect(text).toHaveTextContent(initialValuesMock.description)
  })

  it('should call correctly when handleDelete is passed', () => {
    const handleDelete = jest.fn()
    const initialValue: Props = { ...initialValuesMock, handleDelete }
    setup(initialValue)
    const buttonRemove = screen.getByTestId('input-todo-delete')
    fireEvent.click(buttonRemove)
    expect(handleDelete).toBeCalled()
  })

  it('should complete an item correctly when handleCompleteItem is called', () => {
    setup(initialValuesMock)
    const container = screen.getByTestId('todo-container')
    const text = screen.getByTestId('todo-text')
    fireEvent.click(container)
    waitFor(() => expect(text).toHaveStyle('text-decoration: line-through'))
  })
})
