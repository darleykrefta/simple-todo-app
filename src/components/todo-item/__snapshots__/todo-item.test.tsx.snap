// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component should be valid with last version 1`] = `
<DocumentFragment>
  <div
    class="css-14jp1cp"
  >
    <div
      class="css-1xecl85"
      data-testid="todo-container"
    >
      <p
        class="chakra-text css-0"
        data-testid="todo-text"
      >
        todo 1
      </p>
    </div>
  </div>
</DocumentFragment>
`;
